moquette-0.16/broker/src/main/java/io/moquette/broker/PostOffice.java

```diff
@@ -0,0 +1,671 @@
+     *            the result of the enqueuing operation to session loops.
+     */
+    public RoutingResults internalPublish(MqttPublishMessage msg) {
+        final MqttQoS qos = msg.fixedHeader().qosLevel();
+        final Topic topic = new Topic(msg.variableHeader().topicName());
+        final ByteBuf payload = msg.payload();
+        LOG.info("Sending internal PUBLISH message Topic={}, qos={}", topic, qos);
+
+        final RoutingResults publishResult = publish2Subscribers(payload, topic, qos);
+        LOG.trace("after routed publishes: {}", publishResult);
+
+        if (!msg.fixedHeader().isRetain()) {
+            return publishResult;
+        }
+        if (qos == AT_MOST_ONCE || payload.readableBytes() == 0) {
+            // QoS == 0 && retain => clean old retained
+            retainedRepository.cleanRetained(topic);
+            return publishResult;
+        }
+        retainedRepository.retain(topic, msg);
+        return publishResult;
+    }
+
+    /**
+     * notify MqttConnectMessage after connection established (already pass login).
+     * @param msg
+     */
+    void dispatchConnection(MqttConnectMessage msg) {
+        interceptor.notifyClientConnected(msg);
+    }
+
+    void dispatchDisconnection(String clientId,String userName) {
+        interceptor.notifyClientDisconnected(clientId, userName);
+    }
+
+    void dispatchConnectionLost(String clientId,String userName) {
+        interceptor.notifyClientConnectionLost(clientId, userName);
+    }
+
+    /**
+     * Route the command to the owning SessionEventLoop
+     * */
+    public RouteResult routeCommand(String clientId, String actionDescription, Callable<String> action) {
+        SessionCommand cmd = new SessionCommand(clientId, action);
+        final int targetQueueId = Math.abs(cmd.getSessionId().hashCode()) % this.eventLoops;
+        LOG.debug("Routing cmd [{}] for session [{}] to event processor {}", actionDescription, cmd.getSessionId(), targetQueueId);
+        final FutureTask<String> task = new FutureTask<>(() -> {
+            cmd.execute();
+            cmd.complete();
+            return cmd.getSessionId();
+        });
+        if (Thread.currentThread() == sessionExecutors[targetQueueId]) {
+            SessionEventLoop.executeTask(task);
+            return RouteResult.success(clientId, cmd.completableFuture());
+        }
+        if (this.sessionQueues[targetQueueId].offer(task)) {
+            return RouteResult.success(clientId, cmd.completableFuture());
+        } else {
+            LOG.warn("Session command queue {} is full executing action {}", targetQueueId, actionDescription);
+            return RouteResult.failed(clientId);
+        }
+    }
+
+    public void terminate() {
+        for (Thread processor : sessionExecutors) {
+            processor.interrupt();
+        }
+        for (Thread processor : sessionExecutors) {
+            try {
+                processor.join(5_000);
+            } catch (InterruptedException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)